src/vsr.zig

```diff
@@ -1917,3 +1918,52 @@ const ViewChangeHeadersArray = struct {
         headers.array.appendAssumeCapacity(Headers.dvc_blank(op));
     }
 };
+
+pub const Op = struct {
+    pub fn checkpoint_before(checkpoint: u64) ?u64 {
+        assert(checkpoint_valid(checkpoint));
+
+        if (checkpoint == 0) {
+            return null;
+        }
+
+        if (checkpoint == constants.journal_slot_count - constants.lsm_batch_multiple - 1) {
+            return 0;
+        }
+
+        return checkpoint - (constants.journal_slot_count - constants.lsm_batch_multiple);
+    }
+
+    pub fn checkpoint_after(checkpoint: u64) u64 {
+        assert(checkpoint_valid(checkpoint));
+
+        const checkpoint_next = op: {
+            if (checkpoint == 0) {
+                // First wrap: op_checkpoint_next = 8-2-1 = 5
+                break :op constants.journal_slot_count - constants.lsm_batch_multiple - 1;
+            } else {
+                // Second wrap: op_checkpoint_next = 5+8-2 = 11
+                // Third wrap: op_checkpoint_next = 11+8-2 = 17
+                break :op checkpoint + constants.journal_slot_count - constants.lsm_batch_multiple;
+            }
+        };
+
+        assert((checkpoint_next + 1) % constants.lsm_batch_multiple == 0);
+        assert(checkpoint_valid(checkpoint_next));
+        assert(checkpoint_before(checkpoint_next) == checkpoint);
+
+        return checkpoint_next;
+    }
+
+    pub fn trigger_from_checkpoint(checkpoint_nonzero: u64) u64 {
+        assert(checkpoint_nonzero != 0);
+        assert(checkpoint_valid(checkpoint_nonzero));
+
+        return checkpoint_nonzero + constants.lsm_batch_multiple;
+    }
+
+    pub fn checkpoint_valid(op: u64) bool {
+        return op == 0 or
+            (op + 1) % (constants.journal_slot_count - constants.lsm_batch_multiple) == 0;
```

Comment:
Do we have an off-by-one here?

It feels like where we modulo an index by a count, we don't actually want to convert the index to a count?

Reply:
This "+1" is a shift to the right, to counteract the "-1" applied to the first wrap's checkpoint.

Reply:
Could you help me understand this better? I don't follow.

Reply:
In this diagram, `✓` are each checkpoints. Each is one minus a multiple of `constants.journal_slot_count - constants.lsm_batch_multiple`.

```zig
        /// For a replica with journal_slot_count=8 and lsm_batch_multiple=2:
...
        ///     commit log (ops)           │ write-ahead log (slots)
        ///     0   4   8   2   6   0   4  │ 0---4---
        ///   0 ─────✓·%                   │ 01234✓6%   initial log fill
        ///   1 ───────────✓·%             │ 890✓2%45   first wrap of log
        ///   2 ─────────────────✓·%       │ 6✓8%0123   second wrap of log
        ///   3 ───────────────────────✓·% │ 4%67890✓   third wrap of log
        ///
        /// Legend:
        ///
        ///   ─/✓  op on disk at checkpoint
        ///   ·/%  op in memory at checkpoint
        ///     ✓  op_checkpoint
        ///     %  op_checkpoint's trigger
```

The `+ 1` in `(op + 1) % (constants.journal_slot_count - constants.lsm_batch_multiple)` cancels out the `- 1` that is part of the first wrap's checkpoint (7 in the diagram):

```
    pub fn checkpoint_after_checkpoint(checkpoint: u64) u64 {
        assert(checkpoint_valid(checkpoint));

        const checkpoint_next = op: {
            if (checkpoint == 0) {
                // First wrap: op_checkpoint_next = 8-2-1 = 5
                break :op constants.journal_slot_count - constants.lsm_batch_multiple - 1;
            } else {
```

Since every successive checkpoint after the first wrap adds `constants.journal_slot_count - constants.lsm_batch_multiple`, they all have that `-1` in them also.